e2e/internal/e2e/profile/profile.go

```diff
@@ -0,0 +1,230 @@
+func (User) String() string {
+	return "User"
+}
+
+// Root implements the Profile interface describing a privileged user.
+type Root struct {
+	baseProfile
+}
+
+func (Root) Privileged() bool {
+	return true
+}
+
+func (Root) User(t *testing.T) *user.User {
+	u, err := user.GetPwUID(0)
+	if err != nil {
+		t.Fatalf("failed to retrieve root user information: %s", err)
+	}
+	return u
+}
+
+func (p Root) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (Root) String() string {
+	return "Root"
+}
+
+// Fakeroot implements the Profile interface describing a "fake" root
+// user and makes use of singularity --fakeroot flag where appropriate.
+type Fakeroot struct {
+	baseProfile
+}
+
+func (Fakeroot) Requirements(t *testing.T) {
+	require.UserNamespace(t)
+
+	// check that current user has valid mappings in /etc/subuid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubUIDFile, uint32(e2e.OrigUID())); err != nil {
+		t.Fatalf("fakeroot configuration error: %s", err)
+	}
+
+	// check that current user has valid mappings in /etc/subgid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubGIDFile, uint32(e2e.OrigUID())); err != nil {
+		t.Fatalf("fakeroot configuration error: %s", err)
+	}
+}
+
+func (Fakeroot) Args(cmd []string) []string {
+	commands := []string{
+		"shell",
+		"exec",
+		"run",
+		"test",
+		"instance start",
+		"build",
+	}
+
+	return buildArgs(cmd, commands, "--fakeroot")
+}
+
+func (Fakeroot) User(t *testing.T) *user.User {
+	return getOrigUser(t)
+}
+
+func (p Fakeroot) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (Fakeroot) String() string {
+	return "Fakeroot"
+}
+
+// UserNamespace implemets Profile describing an unprivileged user that
+// operates in a separate user namespace.
+type UserNamespace struct {
+	baseProfile
+}
+
+func (UserNamespace) Args(cmd []string) []string {
+	commands := []string{
+		"shell",
+		"exec",
+		"run",
+		"test",
+		"instance start",
+	}
+
+	return buildArgs(cmd, commands, "--userns")
+}
+
+func (UserNamespace) Requirements(t *testing.T) {
+	require.UserNamespace(t)
+}
+
+func (p UserNamespace) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (UserNamespace) String() string {
+	return "UserNamespace"
+}
+
+// RootUserNamespace implemets Profile describing a privileged user that
+// operates in a separate user namespace.
+type RootUserNamespace struct {
+	UserNamespace
+	Root
+}
+
+func (p RootUserNamespace) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (RootUserNamespace) String() string {
+	return "RootUserNamespace"
+}
+
+// in looks for profile p in the given list of profiles.
+func in(p Profile, profiles []Profile) bool {
+	for _, profile := range profiles {
+		if profile == p {
+			return true
+		}
+	}
+	return false
+}
+
+// buildArgs builds the necessary arguments for singularity command cmd
+// from a list of commands that support the specified flag.
+func buildArgs(cmd []string, commands []string, flag string) []string {
```

Comment:
Overall this looks okay, but I think what you are trying to do is to check whether one slice contains at least one element from the other.
Maybe this func should be changed to smth like `Contains(s, sub []string)`? and then in `Args()` you will do e.g :

```
if Contains() {
return []string{"--fakeroot"}
}
```

Reply:
I think I see what you mean.

This function started off as something different, so I loss some perspective in the process.

This is taking something that already exists as a slice (cmd), which will in most cases contain a single element, but in some cases it will contain two, and maybe more.

Then it takes the list of commands, commands, and looks for cmd in that list. Since cmd is a slice and commands is a slice (not a slice of slices), what the code does is concatenate the elements of cmd and looks for _that_ in the commands slice. If it finds it, it then returns the flag argument a a slice.

All of this is super weird. I'll try the suggested change.